File: data-structures/stack/StackUsingLinkedList.go
// Implementing stack using LinkedList

package main

import "fmt"

type StackNode struct {
	data int
	next *StackNode
}

type Stack struct {
	root *StackNode
}


func (s *Stack) IsEmpty() bool{
	if s.root == nil {
		return true
	}
	return false
}


func (s *Stack) Push(d int) {
	newStackNode := StackNode{data: d,}
	newStackNode.next = s.root
	s.root = &newStackNode
}

func (s *Stack) Pop() {
	if s.IsEmpty() {
		fmt.Println("The Stack is empty")
		return
	}
	rootNode := s.root
	s.root = s.root.next
	popped := rootNode.data
	fmt.Printf("%d Popped from stack\n", popped)
}

func (s *Stack) Peek(){
	if s.IsEmpty() {
		fmt.Println("The stack is empty")
		return
	}
	fmt.Println(s.root.data)
}

func main () {
	s := Stack{}
	s.Push(1)
	s.Push(2)
	s.Push(3)
	s.Push(4)
	s.Push(5)
	s.Peek()
	s.Pop()
	s.Peek()
	fmt.Println(s.IsEmpty())
}

File: python/data_structures/stack/stack_using_linked_list.py
from __future__ import annotations


class StackNode:
    def __init__(self, data: str | None = None, node: StackNode | None = None):
        self.data = data
        self.next = node


class Stack:
    def __init__(self):
        self.head = None

    def create_stack_node(self, data) -> StackNode:
        return StackNode(data)

    @property
    def is_empty(self) -> bool:
        return self.head == None

    def push(self, data: str) -> None:
        new_node = self.create_stack_node(data)
        new_node.next = self.head
        self.head = new_node
        print(f"Inserted {data} to Stack")

    def pop(self) -> None:
        if self.is_empty:
            print("Stack is empty")
            return
        head = self.head
        self.head = head.next
        print(f"{head.data} popped successfully")

    def peak(self) -> None:
        print(f"Peak: {self.head.data}")

    def print_values(self) -> None:
        if self.is_empty:
            print("Stack is empty")
            return

        head = self.head
        while head is not None:
            print(f"{head.data}", end=" ")
            head = head.next
        print()


if __name__ == "__main__":
    stack = Stack()

    # push to stack
    for i in range(0, 6):
        stack.push(str(i))

    # peak
    stack.peak()  # should print 5

    # pop
    stack.pop()  # should return 5

    # print values
    stack.print_values()

    # pop
    stack.pop()  # should return 4

    # peak
    stack.peak()  # should print 3

    stack.print_values()
